File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>


using namespace std;

int N;
std::vector<int> visited;
std::vector<std::vector<int>> v;


void dfs(int n) {
    for (int i = 0; i < v[n].size(); i++) {
        if (!visited[v[n][i]]) {
            visited[v[n][i]] = true;
            dfs(v[n][i]);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    N = computers.size();
    visited = std::vector<int>(N, false);
    
    for (int i = 0; i < N; i++) {
        v.emplace_back(std::vector<int>());
        for (int j = 0; j < computers[i].size(); j++) {
            if (i != j && computers[i][j] == 1) {
                v[i].emplace_back(j);
            }
        }
    }
    
    for (int i = 0; i < N; i++) {
        if (!visited[i]) {
            visited[i] = true;
            answer++;
            dfs(i);
        }
    }
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {

    int[] check;
    
    void dfs(int k, int[][] computers) {
        check[k] = 1;
        for (int i = 0; i < computers.length; i++) {
            if (check[i] == 0 && k != i && computers[k][i] == 1) {
                dfs(i, computers);
            }
        }
    }

    public int solution(int n, int[][] computers) {
        int answer = 0;
        check = new int[computers.length];
        for (int i = 0; i < computers.length; i++) {
            check[i] = 0;
        }
        for (int i = 0; i < computers.length; i++) {
            if (check[i] == 0) {
                dfs(i, computers);
                answer++;
            }
        }
        
        return answer;
    }
}
